proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,621 @@
+account can do lamport transfer to recover these fees. So instead of fees going
+to the validator, these fees go to the **dapp developers**. It will be dapp
+developer's responsibility to advertise the required PRAW fees to its users.
+These fees will depend on the CU requested by the transaction.
+
+## Motivation
+
+Currently, there is no way for dapp developers to enforce appropriate behavior
+and the way their contracts should be used. Bots spamming on dapps make them
+unusable and dapps lose their users/clients because the UX is laggy and
+inefficient. Unsuccessful transactions deny block space to potentially valid
+transactions which reduces activity on the dapp. For dapps like mango markets or
+Openbook increasing fees without any rebates or dynamically based other proposed
+mechanisms will punish potentially thousands of users because of a handful of
+malicious users. Giving dapps authority more control to incentivize proper
+utilization of its contract is the primary motivation for this proposal.
+
+During network congestion, many transactions in the network cannot be processed
+because they all want to write-lock the same accounts. When a write lock on an
+account is taken by a transaction batch no other batch can use this account in
+parallel, so only transactions belonging to a single batch are processed
+correctly and all others are retried again or forwarded to the leader of the
+next slot. With all the forwarding and retrying the validator is choked by the
+transactions write-locking the same accounts and effectively processing valid
+transactions sequentially.
+
+There are multiple accounts of OpenBook, mango markets, etc which are used in
+high-frequency trading. During the event of extreme congestion, we observe that
+specialized trading programs write lock these accounts but never CPI into the
+respective programs unless they can extract profit, effectively starving the
+actual users for access. With current low cluster fees, it incentivizes spammers
+to spam the network with transactions.
+
+Without any proper rebate mechanism for users, Solana fees will increase and it
+will lose the edge for being low fees cluster. For entities like market makers,
+Solana must remain a low-fee cluster as they have a thin profit margin and have
+to quote quite often. The goal of this proposal is to reduce spamming without
+really increasing fees for everyone. Encourage users to create proper
+transactions. Read the cluster state before creating transactions instead of
+spamming. We are motivated to improve the user experience, and user activity and
+keep Solana a low gas fee blockchain. Keeping gas fees low and increasing user
+activity on the chain will help all the communities based on Solana grow. This
+will also help users to gain more control over accounts owned by them.
+
+Note that this proposal does not aim to increase gas fees for all transactions
+or add any kind of congestion control mechanism.
```

Comment:
> add any kind of congestion control mechanism.

Isn't PRAW fee a way of ease accounts' congestion?

Reply:
Not directly at the runtime level. Network congestion mechanism always get confused at very low level. While PRAW fees might be useful to control congestion of few accounts. The control is by discouraging spamming, if dapp is used by many users and network is congested by many valid transactions then PRAW fees cannot help to control the congestion. We can think that prioritization fees are some kind of congestion control mechanism.